### ISSUE TITLE
Collecting coverage under QNX

### ISSUE THREAD

**Issue by latk** [gcovr/gcovr#316]
_Originally posted by @nikhilbagul in https://github.com/gcovr/gcovr/issues/259#issuecomment-500504846_

I am running into a similar issue. I am building and running a UnitTest app for shared libraries made in QT. I am running it on an arm_v8 platform running QNX 7.0

Below are the environment specs to help you understand better: 

Compiler toolchain: 
qmake (3.1) (Qt 5.12.0)
qcc

I followed the steps mentioned here : 
[gcov on QNX](https://www.qnx.com/developers/docs/7.0.0/#com.qnx.doc.neutrino.utilities/topic/g/gcov.html)

After playing around much I am still trying to get my head around why the .gcda files generated on target do not produce a **gcovr** code coverage report when moved back to the host. 

Below is the log for reference: 

```
Filters for --root: (1)
- <_sre.SRE_Pattern object at 0x7facca83d200>
Filters for --filter: (1)
- DirectoryPrefixFilter(\/Users\/nikhil\.bagul\/QtProjects\/VPCluster\/InstrumentCluster\/InstrumentCluster\/DatahandlerUtil\/FixedID\/src\/)
Filters for --exclude: (0)
Filters for --gcov-filter: (1)
- AlwaysMatchFilter()
Filters for --gcov-exclude: (0)
Filters for --exclude-directories: (0)
Scanning directory . for gcda/gcno files...
Found 5 files (and will process 3)
Pool started with 1 threads
Processing file: /Users/nikhil.bagul/QtProjects/VPCluster/InstrumentCluster/InstrumentCluster/DatahandlerUtil/FixedID/src/fixed_id.gcda
Running gcov: 'gcov /Users/nikhil.bagul/QtProjects/VPCluster/InstrumentCluster/InstrumentCluster/DatahandlerUtil/FixedID/src/fixed_id.gcda --branch-counts --branch-probabilities --preserve-paths --object-directory /Users/nikhil.bagul/QtProjects/VPCluster/InstrumentCluster/InstrumentCluster/DatahandlerUtil/FixedID/src' in '/var/folders/_2/pxhj8y854g54d1bzsqkbt0jj8_vp25/T/tmpeNuONP'
Processing file: /Users/nikhil.bagul/QtProjects/VPCluster/InstrumentCluster/InstrumentCluster/DatahandlerUtil/FixedID/src/moc_fixed_id.gcda
Running gcov: 'gcov /Users/nikhil.bagul/QtProjects/VPCluster/InstrumentCluster/InstrumentCluster/DatahandlerUtil/FixedID/src/moc_fixed_id.gcda --branch-counts --branch-probabilities --preserve-paths --object-directory /Users/nikhil.bagul/QtProjects/VPCluster/InstrumentCluster/InstrumentCluster/DatahandlerUtil/FixedID/src' in '/var/folders/_2/pxhj8y854g54d1bzsqkbt0jj8_vp25/T/tmpeNuONP'
Processing file: /Users/nikhil.bagul/QtProjects/VPCluster/InstrumentCluster/InstrumentCluster/DatahandlerUtil/FixedID/src/moc_fixed_values.gcno
Running gcov: 'gcov /Users/nikhil.bagul/QtProjects/VPCluster/InstrumentCluster/InstrumentCluster/DatahandlerUtil/FixedID/src/moc_fixed_values.gcno --branch-counts --branch-probabilities --preserve-paths --object-directory /Users/nikhil.bagul/QtProjects/VPCluster/InstrumentCluster/InstrumentCluster/DatahandlerUtil/FixedID/src' in '/var/folders/_2/pxhj8y854g54d1bzsqkbt0jj8_vp25/T/tmpeNuONP'
Gathered coveraged data for 0 files
```

The generated coverage report is empty.

I have also tried running gcov on an individual source file to run into the following issue: 

Unexpected version: *405.
Invalid .gcno File!

I double checked my paths and environment variables on the target platform with no luck, I am running out of ideas here. If you had any luck on generating coverage in your cross-platform venture it will be super helpful if you can share some wisdom. 

Thanks bud!

Also attaching the Makefile for reference.
[Makefile.txt](https://github.com/gcovr/gcovr/files/3273139/Makefile.txt)

**Comment by latk** [comment]
Hi @nikhilbagul, the version of the gcov tool that you use must match the compiler. When cross-compiling, you cannot just use the standard `gcov` executable that's provided by whatever GCC version you have installed.

The QNX docs you linked mention that the gcov variants are named after the target platform, like `ntoaarch64-gcov`.

You can tell gcovr to use a specific gcov tool:

- by setting the `GCOV` environment variable
- by setting the `--gcov-executable` command line argument

Can you check if using the correct gcov variant lets you process an individual gcda file? If so, gcovr should work as well. If there are further issues they likely have to do with gcovr filters or your directory layout.

**Comment by nikhilbagul** [comment]
I am sorry, this may sound amateur, but can you guide on how to point to the correct gcov or set GCOV or gcovr to use the right variant? I am trying to find some documentation for the same

**Comment by latk** [comment]
I have no idea about QNX. But under the assumption that you are targeting ARMv8, you would probably have to run gcovr like:

``` sh
gcovr --gcov-executable=ntoaarch64-gcov OTHER_OPTIONS
```

In your terminal shell (Linux, mac, not Windows) you can set environment variables like

``` sh
export GCOV=ntoaarch64-gcov
```

The `export` makes the variable visible to any programs that you launch from the shell.

If you are running gcovr via some tool and not directly via a shell, I can't help with such third party tools.

Gcovr's --gcov-executable option is documented [here](https://www.gcovr.com/en/stable/guide.html#cmdoption-gcovr-gcov-executable).

**Comment by nikhilbagul** [comment]
Hey buddy, I was able to generate the report. Thanks to you, It was a matter of pointing to the correct **gcov** executable. 
But the online documentation for QNX lacked the necessary info on where to find it. The options mentioned in the link I mentioned above don't work. 

`Target platform	gcov_variant
ARMv7	ntoarmv7-gcov
AArch64	ntoaarch64-gcov
x86	ntox86-gcov
x86 64-bit	ntox86_64-gcov`

Upon further digging I found the correct executable buried deep in the QNX SDP on my host machine. 

Hence executing the command - 

`gcovr --gcov-executable=/qnx700/host/darwin/x86_64/usr/bin/aarch64-unknown-nto-qnx7.0.0-gcov -r .`

was able to generate the desired reports. 
Thanks for all the help. Closing out the ticket for now. :)
